### ISSUE TITLE
Transaction Details and Signing

### ISSUE THREAD

**Issue by Shadowfiend** [tallycash/extension#369]
[Figma prototype](https://www.figma.com/proto/q8gfu9XTbYBF07KMnDCtsD/UI---Dark?page-id=3582%3A106079&node-id=3582%3A106124&viewport=241%2C48%2C0.1&scaling=scale-down&starting-point-node-id=3582%3A106201&show-proto-sidebar=1)

### Context

The transaction signing flow is entered from one of a few places:
- Global Send button at the top of the account tab.
- Asset-specific Send button on the account tab's asset list or from an asset's details page.
- A dApp request for signing, in which the wallet will be popped up on the sign transaction page with the signing details provided by the dApp.
- A swap transaction submission, from the swap tab.

Transaction signing displays information about the transaction, depending on transaction type:
- Recipient of the transaction. An address, with the possibility of showing richer information when available (out of scope for this issue).
- ETH that will be sent to the recipient with the transaction.
- Transaction data (possibly under the Advanced tab).
- Selected network fees, as well as the ability to update them.

Certain transactions can provide additional details, though this may be out of scope for day 1. More details on this in [the Sign Transactions](https://www.figma.com/file/q8gfu9XTbYBF07KMnDCtsD/UI---Dark?node-id=1917%3A64142) screens in Figma.

### Currently implemented

- With `HIDE_SEND_BUTTON` set to `true`, clicking on an asset will display a `Send` button that will transition to the send screen. This is the only access point to existing sign transaction work.
  - Clicking the asset dropdown on the “Send Asset” page and selecting the only asset will enable the “Send” button.
  - Network fee/speed section is 
  - Clicking the “Send” button again will transition to the signing screen.
- Basic structure of the transaction to be signed exists, as do variants on transaction signatures.

### To Be Implemented

- Fee selector UI. This should use the `transaction-construction` slice's `gasEstimates` field as input.
- `chain/index.ts`'s `pollBlockPrices` should be updated to use the `BaseService`'s constructor's alarms functionality instead of `setTimeout`.
- Plugging `Send`, `SignTransaction`, etc into the `transaction-construction` slice's `transactionRequest` and `updateTransactionOptions` thunk. This includes hooking up the fee selector UI to update transaction options for the current transaction.
- Augmenting the `transaction-construction` slice to request signature of a completed transaction request. This is currently implemented in `main.ts`[^1], but not hooked up anywhere. It also runs every time transaction options are requested, whereas it should run once, when the user hits “Sign” on the `SignTransaction` page.
- Allowing the possibility of an unlock screen after hitting Sign but before signing (if the wallet is locked). This will probably require some collab with #423 .
- Reacting to the transaction broadcast once the transaction is signed; currently hooked up in `main.ts`[^2], but the UI does not react and it hasn't been confirmed that the transaction is added to the wallet's transactions list as pending.

### To Be Resolved

- What is in Details vs Advanced for initial release? Currently these have minimal information that may not benefit from the tabbed interface.
- Once the user finishes signing and the transaction is broadcast, what's the next step? Is it always the same no matter where the user entered the sign transaction flow?

[^1]: https://github.com/tallycash/tally-extension/blob/191c03a5d31e33ebec815d6e177f5ffc3de30bba/background/main.ts#L282-L321
[^2]: https://github.com/tallycash/tally-extension/blob/191c03a5d31e33ebec815d6e177f5ffc3de30bba/background/main.ts#L385-L403

**Comment by VladUXUI** [comment]
> What is in Details vs Advanced for initial release? Currently these have minimal information that may not benefit from the tabbed interface.

Totally agreed, for first release we can remove the Tabs. Will mock up shortly.

>Once the user finishes signing and the transaction is broadcast, what's the next step? Is it always the same no matter where the user entered the sign transaction flow?

Users should be taken to the Activity tab of that wallet with a pop-over of a snackbar confirming their broadcast. This might change down the road with the addition of Earn and other cases i can't really foresee now

**Comment by Shadowfiend** [comment]
> with a pop-over of a snackbar confirming their broadcast

Is there a node in Figma you can link to for this? Didn't spot one when I was poking around but there's a lot in there :grin:

**Comment by VladUXUI** [comment]
Can be seen here. But it's not a specific one 
https://www.figma.com/file/q8gfu9XTbYBF07KMnDCtsD/?node-id=570%3A0

Also adding a screenshot.
![image](https://user-images.githubusercontent.com/80846886/143065186-80868ca2-b7ba-46e8-9349-9696c083bcbe.png)

This component is not yet built as far as i know

**Comment by wojciech-turek** [comment]
- `Fee selector UI. This should use the transaction-construction slice's gasEstimates field as input.` - Created the UI for it and created a selector for gasEstimates however its always null, not sure when `pollBlockPrices` trigger.
- `chain/index.ts's pollBlockPrices should be updated to use the BaseService's constructor's alarms functionality instead of setTimeout.` - https://github.com/tallycash/tally-extension/pull/471/files in `background/services/chain/index.ts` I have left out commented out code because I am not sure I am understanding it correctly, should I create a new alarm or use a specific alarm that already exist? Haven't used that functionality before.

@Shadowfiend

**Comment by Shadowfiend** [comment]
I believe you'll need a blocknative API key in your env for the prices. Left notes on the alarms stuff in the PR!

**Comment by wojciech-turek** [comment]
@Shadowfiend

> Plugging Send, SignTransaction, etc into the transaction-construction slice's transactionRequest and updateTransactionOptions thunk. This includes hooking up the fee selector UI to update transaction options for the current transaction.

My understanding is that the flow is as follows: 
- User selects token, amount and gasOption and click `Send` which triggers `transactionOptions` dispatch.
- User is taken to sign page (@VladUXUI can I please ask which design exactly) where user can click `Sign`, its like a final step before actually sending anything.
- Sign button triggers `transactionRequest` dispatch (not existing yet) which creates `transaction` which is later signed by keyringService.signTransaction() method.
- Transaction options are cleared

**Comment by Shadowfiend** [comment]
We should be able to update the transaction options as we go imo, but I don't remember if that's true in practice.

**Comment by wojciech-turek** [comment]
@Shadowfiend I have added asyncBackroundThunks for creating a transaction and signing it [here](https://github.com/tallycash/tally-extension/pull/471/files#diff-613eb2ecd9deba7bd12181aa94f29ae0c54a8685e9cf6673e3cf77a529fcfb7f).

I would like to ask for your help regarding: 
 - How to properly call chain service to get a gasLimit from the redux-slice file, and should that be called in Send component since there is a gasLimit input? I was thinking something like: 
 > transactionData.gasLimit = await chainService.estimateGasLimit(getEthereumNetwork(), transaction)

-  Is gasPrice coming from chains service in gwei? What is the difference between that and baseFeePerGas from blocknative?
- The price (not sure if price === gasPrice from chainservice) is also coming from gasEstimates from blocknative so do I even need to make that call?
  > transactionData.gasPrice = await chainService.pollingProviders.ethereum.getGasPrice()
  
- How to call keyring service correctly to sign transaction from the redux-slice? (if same as above for chain service please skip this question)
> const signedTx = keyringService.signTransaction(transaction.from, transaction)

- Do I push the signed tx as pending somewhere or is that done automatically, I couldn't find.

**Comment by Shadowfiend** [comment]
> gasLimit

This is already automatically called when `updateTransactionOptions` runs[^1]. We should lean into that implementation, which runs every time transaction details are updated by the user, and wire up the UI to read its current state from the transaction options.

> gasPrice

gasPrice is irrelevant for us at the moment, as it's only applicable to legacy EVM transactions (pre-EIP1559)[^2]. We should ignore Blocknative's value as well, and use their EIP-1559 fields for maxFeePerGas and maxPriorityFeePerGas.

> How to call keyring service correctly to sign transaction from the redux-slice?

This is a little trickier. The call itself should be the same one that currently runs every time `updateTransactionOptions` runs[^3], but it should be updated to run only once when the Sign button is clicked. I expect we'll want a new background async thunk, a new event in the construction slice, and a new hookup in main from that event to sign and then broadcast.

> Do I push the signed tx as pending somewhere

Not directly, but `ChainService` has a `broadcastSignedTransaction` method that takes care of both broadcasting it and adding it to local tracking[^4].

[^1]: https://github.com/tallycash/tally-extension/blob/ea2a322ab274f8a751afaf29e4c8d23f950c4e6d/background/main.ts#L282
[^2]: https://github.com/tallycash/tally-extension/blob/310a812145f4a29af5541058154d4eb38819a602/background/networks.ts#L105-L115 https://github.com/tallycash/tally-extension/blob/310a812145f4a29af5541058154d4eb38819a602/background/networks.ts#L130-L140
[^3]: https://github.com/tallycash/tally-extension/blob/310a812145f4a29af5541058154d4eb38819a602/background/main.ts#L319
[^4]: https://github.com/tallycash/tally-extension/blob/310a812145f4a29af5541058154d4eb38819a602/background/services/chain/index.ts#L353

**Comment by wojciech-turek** [comment]
> Totally agreed, for first release we can remove the Tabs. Will mock up shortly.

@VladUXUI, can you please link me to it, I cant find the updated version.

**Comment by VladUXUI** [comment]
Here is a preview. I only did this for one of them. Let me know if you need me to do for both
![image](https://user-images.githubusercontent.com/80846886/145216883-16316493-7f22-4d99-87ec-b94ccdef9abc.png)

**Comment by mhluongo** [comment]
Mostly closed by #1263. The rest of this is tracked elsewhere and will be iterated on :)
